### PR TITLE
Add a `close` command

### PR DESCRIPTION
**Pull request by joneshf**
We want to be able to close a PR from `spr`.

This follows a lot of what happens in the other commands: we grab the
commit, find its related PR, close it, and delete any branches.

We also give it a `--all` option so that all PRs in a stack can be
closed at once.

Closes: https://github.com/getcord/spr/issues/63

Test Plan:
There are three scenarios we want to check: closing a PR based on the
trunk branch, closing a PR based on a non-trunk branch, and closing a
stack of PRs.

## Trunk branch

1. Create a PR with `spr diff`.
1. Run `spr close`.
    - [ ] The PR should be closed
    - [ ] The head branch should be deleted on GitHub.
    - [ ] The commit message should no longer have the "Pull Request" or
        "Reviewed By" sections.

## Non-trunk branch

1. Create one PR with `spr diff`.
1. Create a second PR with `spr diff`.
1. Run `spr close` on the second PR.
    - [ ] The PR should be closed
    - [ ] The head branch should be deleted on GitHub.
    - [ ] The base branch should also be deleted on GitHub.
    - [ ] The commit message should no longer have the "Pull Request" or
        "Reviewed By" sections.

## Stack of PRs

1. Create one PR with `spr diff`.
1. Create a second PR with `spr diff`.
1. Run `spr close --all` on the second PR.
    - [ ] Each PR should be closed
    - [ ] Each head branch should be deleted on GitHub.
    - [ ] Each base branch should also be deleted on GitHub.
    - [ ] Each commit message should no longer have the "Pull Request"
        or "Reviewed By" sections.

### CODE REVIEW

File: src/commands/close.rs
Comment (sven-of-cord) [inline src/commands/close.rs:76]: At this point, it would be nice if spr rewrote the local commit's message to remove the "Pull Request:" section. Because then you could open a new PR with "spr diff" if you wanted, without the need to change the commit message yourself. Also, it would just be "more correct", you know.

I think it should be easy. Haven't tried it, but something along the lines of
```

```

Comment (sven-of-cord) [inline src/commands/close.rs:76]: At this point it would be nice to rewrite the local commit's message and remove the "Pull Request:" section. That would mean that you could open a new PR immediately with `spr diff`. Also it's just "more correct", you know...

I haven't tried it, but it should be easy, something along the lines of

```
prepared_commit.message.remove(&MessageSection::PullRequest);
drop(prepared_commit);
git.rewrite_commit_messages(prepared_commits.as_mut_slice(), None)?;
```

File: src/commands/close.rs
Comment (sven-of-cord) [inline src/commands/close.rs:78]: I just landed #87 - you will have to rebase your PR and change `async::process` to `tokio::process`. See https://github.com/getcord/spr/pull/87/files#diff-e0cf2851ba3ab9d47423403ab3d5b29812959994245d5ca612a06b92629d608fL361-L388 for the details.

**Review by sven-of-cord — COMMENTED** [review]
This is a nice little addition. Amazing work!

Thanks for doing this. I have seen your other pull request, but I probably won't have time looking into that for a few days. Don't worry though, I'm sure we'll get that landed, too, eventually. I am starting with this PR because it's much quicker to review.

I left a couple of inline comments.

One more idea for `spr close` would be to add an `--all` option for closing all PRs in the local branch. (Much like `spr diff --all`)

Thanks again!

File: src/commands/close.rs
Comment (joneshf) [reply src/commands/close.rs:76]: Sounds good! Lemme try and implement that.

File: src/commands/close.rs
Comment (joneshf) [reply src/commands/close.rs:78]: Thanks for the heads up!

**Review by joneshf — COMMENTED** [review]
> Thanks for doing this. I have seen your other pull request, but I probably won't have time looking into that for a few days. Don't worry though, I'm sure we'll get that landed, too, eventually. I am starting with this PR because it's much quicker to review.

No worries! Take your time.

I can also split that PR into smaller PRs that could be easier to review. I mean, that's one of the promises of this whole stacked diff workflow. I just don't think I can stack them very easily at the moment (since I'm coming from a fork and it doesn't really work with the current algorithm).

> One more idea for `spr close` would be to add an `--all` option for closing all PRs in the local branch. (Much like `spr diff --all`)

Sounds good!

**Comment by joneshf** [comment]
Yeah, I'm just making it worse at this point. Might need to either: use actual `git` commands to rewrite the way it's supposed to be, or just open a new PR. Either way, going to deal with this later.

**Comment by joneshf** [comment]
Aight. I think it's sorted now. For posterity, I ended up doing:
```Console
# Push the original commit to the branch
$ git push --force-with-lease origin 2d31f24:spr/joneshf/add-a-close-command
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
To github.com:joneshf/spr.git
 + 3a1c30b...2d31f24 2d31f24 -> spr/joneshf/add-a-close-command (forced update)
# Remove the two branches created by accident.
$ git push origin :spr/joneshf/master.add-a-close-command
To github.com:joneshf/spr.git
 - [deleted]         spr/joneshf/master.add-a-close-command
$ git push origin :spr/joneshf/master.add-a-close-command-1 
To github.com:joneshf/spr.git
 - [deleted]         spr/joneshf/master.add-a-close-command-1
# Update now that everything should be sorted
$ spr diff
25138c9 Add a `close` command
  #️⃣   Pull Request #109: https://github.com/getcord/spr/pull/109
Message (leave empty to abort): Rebase atop `master` branch
  ⚾  Commit was rebased - updating Pull Request #109
```

I _think_ that what I should've done originally was:
```Console
# Get the trunk branch up to date
$ git checkout master
$ git pull upstream --rebase
# Fix any conflicts.
# Update the ref to the remote trunk branch.
# I think this was the part I was missing.
$ git push origin
# Switch to the commit for this PR.
$ git checkout 2d31f24
# Rebase on top of the updated trunk
$ git rebase --onto master
# Update the PR
$ spr diff
```

Might be something worth discussing the fork PR or issue, rather than here. But I expected that the upstream trunk branch would be the one that was important, not the origin trunk branch. There's a good chance I missed a change in that PR though, so I wouldn't worry about it too much.

In any case, that's a lot of steps and knowledge to remember. Might make sense to have a command that does a lot of this stuff when working from a fork (or just in general).

File: src/commands/close.rs
Comment (joneshf) [reply src/commands/close.rs:120]: So, I'm very much a Rust novice and what you showed worked. But I'm curious, why do we need the `drop`? From what I've read about `drop` and how it works, it seems like since the `prepared_commit` goes out of scope, it shouldn't be a big deal. But I recognize that I'm probably missing some stuff here.

The main reason I ask is that after adding the `--all` option and extracting a function (to make things a little more understandable/similar) the `drop` seemed a little superfluous and I left it out. But can easily add it back in if that's the right way.

**Comment by joneshf** [comment]
Okay, I think I've implemented all the suggestions. Mind giving it another look whenever you get a minute? More than happy to follow up on whatever else.

File: src/commands/close.rs
Comment (sven-of-cord) [inline src/commands/close.rs:117]: As the reader of this code, I now wonder what the type of `closed` is, and if we should do anything with it. We pass it out of the `Ok`-branch of this `match`, but then it's discarded.

I checked out the branch and saw that `closed` is `()`.

I suggest to change the `Ok` branch like this: `Ok(()) => (),`, then it's self explanatory.

My initial reaction was to think that I'd like `if let Err(error) = result` more, but now thinking about it, that also just ignores whatever non-error value `result` could have. So now I like using `match` here. If you replace `closed` in there with `()`, then the code makes it clear on the first sight: `result` can be one of two things, either an empty Ok (`Ok(())`), or an error.

I am still very much in the learning phase of Rust. This comment is me thinking out loud.

File: src/commands/close.rs
Comment (sven-of-cord) [reply src/commands/close.rs:120]: I can't say with confidence either. Did you try it without the `drop`? Maybe it works.

My thinking was that `prepared_commit` borrows something from within `prepared_commits`, so I wanted to end that before using `prepared_commits`. But maybe Rust figures that out without `drop`?

Not that we need to know now, with the changes above the `drop` isn't needed anymore anyway.

**Review by sven-of-cord — COMMENTED** [review]
This is looking very good.

There is this one one-line change I suggested in the inline comment. Other than that, I think this is ready to go! Yay!

File: src/commands/close.rs
Comment (sven-of-cord) [inline src/commands/close.rs:88]:
```suggestion
    let pull_request = gh.clone().get_pull_request(pull_request_number).await?;
```
One more rebase required, sorry!

File: src/commands/close.rs
Comment (joneshf) [reply src/commands/close.rs:45]: I'm not entirely sure about a Rust-way. But before this PR, there are three commands with an `--all` option (`amend`, `diff`, and `format`), this PR adds a fourth, and there might be a fifth–depending on how https://github.com/getcord/spr/issues/107 goes. I assume at some point, the "process the `--all` option" algorithm will replace these snippets. Maybe when it does that, something idiomatic will show itself.

File: src/commands/close.rs
Comment (joneshf) [reply src/commands/close.rs:88]: No worries! Gives me a chance to try out the ["Probably the proper approach" section](https://github.com/getcord/spr/pull/109#issuecomment-1179429809) and see if it works.

File: src/commands/close.rs
Comment (joneshf) [reply src/commands/close.rs:117]: This all makes sense!

I tell you, as someone with more experience in purely functional languages (PureScript, Haskell, elm, etc.) I expected that if this didn't evaluate to `()` it would be either a compile error or at least a lint error for an unused value. Similar to how if you were to write like `123;` on a line all by itself I would also expect it to be a compile/lint error. I see now that that's not the case, and that's probably fine. Just not what I expected.

I think what you suggest makes sense, though. Lemme update this.

File: src/commands/close.rs
Comment (joneshf) [reply src/commands/close.rs:120]: Yeah, it did work without the `drop`. I just assumed it was one of those things that type checks fine, but could be a potential issue for some reason I'm not familiar with.

**Review by joneshf — COMMENTED** [review]
Alright, cool! Lemme try to update. Finger's crossed I don't break it again 😅.

**Comment by joneshf** [comment]
Alright, ready for another pass. Lemme know if I missed anything!

File: src/commands/close.rs
Comment (joneshf) [reply src/commands/close.rs:120]: Sorry, just had a thought that was related to the original comment:

> At this point it would be nice to rewrite the local commit's message and remove the "Pull Request:" section. That would mean that you could open a new PR immediately with `spr diff`. Also it's just "more correct", you know...

What about the "Reviewers" and "Reviewed By" sections? If the idea is that the "Pull Request" section is no longer part of the commit, then it seems just as reasonable to remove the "Reviewed By" section.

The "Reviewers" section I'm not too sure about. On the on hand, you likely need to change what's in the commit if it's being closed, so you might need a different set of reviewers. Plus, choosing reviewers doesn't work for people without write access to the repo. So it'd cause the next `spr diff` to fail for sure. On the other hand, you can set the reviewers yourself irrespective of the PR.

What's your take on each of these sections?

Comment (sven-of-cord) [reply src/commands/close.rs:120]: Good points. Yes, "Reviewed By" should be removed.

I would leave the "Reviewers" untouched, so they act as the default choice if you resubmit. You can edit the commit message if you want to change that choice. I don't have strong opinions on this one, if you strongly feel "Reviewers" should go, I wouldn't oppose that either.

With regards to the "Reviewers" for people without write access: we should fix that in spr diff, so that it's merely a warning. If you put in a "Reviewers" section, you expect spr to set those for the PR, so if it can't, it should warn you about it, but it shouldn't appear as a failure.

File: src/commands/close.rs
Comment (sven-of-cord) [reply src/commands/close.rs:45]: Sounds good.

File: src/commands/close.rs
Comment (sven-of-cord) [reply src/commands/close.rs]: There is a way in Rust to mark functions such that its return value must not be ignored. Or maybe it was types you mark. I did notice that you can't just ignore `Result<_, _>`, because that would mean to silently ignore errors potentially, which shouldn't happen. Don't remember from the top of my head how it's done, maybe some `#[]`-style directive.

But by default, Rust doesn't complain. I think I'm okay with this approach. Only forces you to write boilerplate to explicitly ignore something, if that something normally shouldn't be ignored.

**Review by sven-of-cord — APPROVED** [review]
I am happy at this point. But by all means, feel free to make the additional change for taking out "Reviewed by".

If I'm not mistaken, my "Approve" on the PR doesn't help you much anyway, because you can't merge. Anyway, it's the workflow I'm used to at work: approve when issues have been solved, sometimes with small optional changes still being discussed, and trust the author with them :wink:

File: src/commands/close.rs
Comment (joneshf) [reply src/commands/close.rs:120]: That makes sense.

Also, if you follow my line of reasoning further:

> On the on hand, you likely need to change what's in the commit if it's being closed, so you might need a different set of reviewers

you end up thinking that maybe the "Summary" and "Test Plan" sections should be removed as well. But that'd be a bit much, I think. So yeah, removing the "Reviewers" section is probably unwarranted.

**Comment by joneshf** [comment]
Thanks for reviewing/merging this!

**Pull request merged**
